File: lib/src/db/entities/user.entity.ts
import { IsBoolean, IsEmail } from "class-validator";
import { BaseEntity } from "../../common/base/base.entity";
import { IUser, UserStatus } from "@traceo/types";
import { Column, Entity, OneToMany, PrimaryGeneratedColumn } from "typeorm";
import { Incident } from "./incident.entity";

@Entity()
export class User extends BaseEntity implements IUser {
  @PrimaryGeneratedColumn("uuid")
  id?: string;

  @Column({ nullable: false, type: "varchar" })
  name: string;

  @Column({ unique: false, type: "varchar" })
  username: string;

  @Column({ type: "varchar", nullable: true })
  @IsEmail()
  email: string;

  @Column({ nullable: true, type: "varchar" })
  @IsEmail()
  gravatar: string;

  @Column({ select: false, nullable: false, type: "varchar", unique: false })
  password: string;

  @Column({ nullable: false })
  status: UserStatus;

  @Column({ nullable: false, default: false, name: "is_admin" })
  @IsBoolean()
  isAdmin: boolean;

  // incidents assigned to this user
  @OneToMany(() => Incident, (incident) => incident.assigned)
  incidents: Incident[];

  @Column({
    nullable: false,
    type: "boolean",
    default: false,
    name: "is_password_updated"
  })
  isPasswordUpdated: boolean;

  @Column({
    type: "bigint",
    nullable: true,
    name: "last_active_at"
  })
  lastActiveAt?: number;
}

File: server/traceo-common/traceo-common-jpa/src/main/java/org/traceo/common/jpa/entities/UserEntity.java
package org.traceo.common.jpa.entities;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.traceo.common.jpa.base.BaseEntity;
import org.traceo.common.transport.dto.api.UserDto;
import org.traceo.common.transport.enums.UserStatus;


@Entity
@Table(name = "traceo_user")
@Getter @Setter
public class UserEntity extends BaseEntity {

    private String name;
    private String username;

    @Column(unique = true)
    private String email;
    private String password;

    private String gravatar;
    private UserStatus status = UserStatus.ACTIVE;

    @Column(name = "is_admin")
    private boolean isAdmin;

    @Column(name = "is_password_updated")
    private boolean isPasswordUpdated;

    @Column(name = "last_active_at")
    private Long lastActiveAt;

    public static UserDto mapToModel(UserEntity entity) {
        UserDto dto = new UserDto();
        dto.setName(entity.getName());
        dto.setAdmin(entity.isAdmin());
        dto.setStatus(entity.getStatus());
        dto.setEmail(entity.getEmail());
        dto.setUsername(entity.getUsername());
        dto.setGravatar(entity.getGravatar());
        return dto;
    }
}
